cli: fix timestamp validation

eggert opened the pull request:
cli/src/cli-cmd-parser.c's config_parse was calling strftime with a struct tm that may have tm_year that is out of range for strftime, yielding unspecified and/or undefined behavior. Fix this by using mktime instead of strftime, as mktime's behavior is well-defined for out-of-range values.  This also fixes a portability issue with strftime %s and time zones.

Fixes: #4355

eggert committed:
cli: fix timestamp validation

cli/src/cli-cmd-parser.c's config_parse was calling strftime
with a struct tm that may have tm_year that is out of range
for strftime, yielding unspecified and/or undefined behavior.
Fix this by using mktime instead of strftime, as mktime's
behavior is well-defined for out-of-range values.  This also
fixes a portability issue with strftime %s and time zones.

Fixes: #4355

Signed-off-by: Paul Eggert <eggert@cs.ucla.edu>

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
Can one of the admins verify this patch?

gluster-ant commented:
CLANG-FORMAT FAILURE:
Before merging the patch, this diff needs to be considered for passing clang-format

```diff --git a/cli/src/cli-cmd-parser.c b/cli/src/cli-cmd-parser.c
index e39f60ed8..ef12ab894 100644
--- a/cli/src/cli-cmd-parser.c
+++ b/cli/src/cli-cmd-parser.c
@@ -2594,7 +2594,7 @@ config_parse(const char **words, int wordcount, dict_t *dict, unsigned cmdi,
                 memset(&checkpoint_time, 0, sizeof(struct tm));
                 ret_chkpt = strptime(append_str, "%Y-%m-%d %H:%M:%S",
                                      &checkpoint_time);
-                checkpoint_sec = mktime (&checkpoint_time);
+                checkpoint_sec = mktime(&checkpoint_time);
 
                 if (ret_chkpt == NULL || *ret_chkpt != '\0' ||
                     checkpoint_time.tm_mday == 0) {
@@ -2610,7 +2610,7 @@ config_parse(const char **words, int wordcount, dict_t *dict, unsigned cmdi,
                     ret = -1;
                     goto out;
                 }
-                snprintf(append_str, 300, "%lld", (long long) checkpoint_sec);
+                snprintf(append_str, 300, "%lld", (long long)checkpoint_sec);
             }
 
             ret = dict_set_dynstr(dict, "op_value", append_str);
```

eggert committed:
cli: pass clang-format check

Adjust patch to pass clang-format.

Fixes: #4355

Signed-off-by: Paul Eggert <eggert@cs.ucla.edu>

eggert committed:
cli: fix off-by-one-hour error if DST

Don't tell mktime that standard time is in effect.
This fixes a bug where the timestamp is off by
an hour when daylight saving time is in effect.
(The issue is also present in the devel branch,
which uses strftime %s and which therefore
has the same bug.)

eggert commented:
I found an extra bug in the code while re-reviewing my patch: there's an off-by-an-hour bug when daylight saving is in effect. This bug is also present in the original unmodified code in the devel branch, as strftime %s has the same issue with tm_isdst that mktime does. I pushed commit  444a9497a30574accdc8b2055eee1e81ce3fdc9d to fix this extra bug.

amarts commented:
/run regression

gluster-ant commented:
1 test(s) failed 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t

0 test(s) generated core 


1 test(s) needed retry 
./tests/00-geo-rep/00-georep-verify-non-root-setup.t
https://build.gluster.org/job/gh_centos7-regression/3423/
